symtabAPI/src/emitElf.C

```diff
@@ -684,7 +684,17 @@ bool emitElf<ElfTypes>::driver(std::string fName) {
             // Clear the PLT type; use PROGBITS
             newshdr->sh_type = SHT_PROGBITS;
         }
-
+        if (library_adjust > 0 &&
+                (strcmp(name, ".init_array") == 0 || strcmp(name, ".fini_array") == 0 ||
+                 strcmp(name, "__libc_subfreeres") == 0 || strcmp(name, "__libc_atexit") == 0 ||
+                 strcmp(name, "__libc_thread_subfreeres") == 0 || strcmp(name, "__libc_IO_vtables") == 0)) {
+            for (int off = 0; off < newdata->d_size; off += sizeof(void*)) {
```

Comment:
Unfortunately, this has a few bits of undefined behavior. The signedness and size of `off` should be the same as `val` since `uint64_t` isn't guaranteed to be the same size as `size_t` which could lead to aliasing; and using a C-style cast to treat `loc` like a `uint64_t*` breaks the aliasing rules.

Here is a version that is standard-conforming:

	for(std::size_t off = 0; off < newdata->d_size; off += sizeof(void*)) {
		char *loc = static_cast<char*>(newdata->d_buf) + off;
		size_t val{};
		memcpy(&val, loc, sizeof(val));
		if(val == 0) continue;
		val += library_adjust;
		memcpy(loc, &val, sizeof(val));
	}

The calls to `memcpy` are required to not break the aliasing rules. gcc and clang remove them even at -O0, so there is no performance penalty here. This will require including `<cstring>`, though.